ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java

```diff
@@ -184,45 +190,107 @@ public Void visitLikePredicate(final LikePredicate node, final TypeContext conte
     @Override
     public Void visitFunctionCall(final FunctionCall node, final TypeContext context) {
       final FunctionName functionName = node.getName();
-
-      // this context gets updated as we process non lambda arguments
-      final TypeContext currentTypeContext = context.getCopy();
-
-      final List<SqlArgument> argumentTypes = new ArrayList<>();
+      final List<SqlArgument> argumentSchemas = new ArrayList<>();
       final List<TypeContext> typeContextsForChildren = new ArrayList<>();
-      final boolean hasLambda = node.hasLambdaFunctionCallArguments();
-      for (final Expression argExpr : node.getArguments()) {
-        final TypeContext childContext = TypeContextUtil.contextForExpression(
-            argExpr, context, currentTypeContext
-        );
-        typeContextsForChildren.add(childContext);
-
-        // pass a copy of the context to the type checker so that type checking in one
-        // expression subtree doesn't interfere with type checking in another one
-        final SqlType resolvedArgType =
-            expressionTypeManager.getExpressionSqlType(argExpr, childContext.getCopy());
-
-        if (argExpr instanceof LambdaFunctionCall) {
-          argumentTypes.add(
+      final List<SqlArgument> argTypesForFirstPass = new ArrayList<>();
+      final List<Expression> arguments = node.getArguments();
+
+      for (final Expression expression : arguments) {
+        if (expression instanceof LambdaFunctionCall) {
+          argTypesForFirstPass.add(
               SqlArgument.of(
-                  SqlLambda.of(currentTypeContext.getLambdaInputTypes(), resolvedArgType)));
+                  SqlLambdaResolved.of(((LambdaFunctionCall) expression).getArguments().size())))
+          ;
         } else {
-          argumentTypes.add(SqlArgument.of(resolvedArgType));
-          // for lambdas - we save the type information to resolve the lambda generics
-          if (hasLambda) {
-            currentTypeContext.visitType(resolvedArgType);
-          }
+          final TypeContext childContext = context.getCopy();
+          final SqlType resolvedArgType = expressionTypeManager.getExpressionSqlType(
+              expression,
+              childContext
+          );
+          argTypesForFirstPass.add(SqlArgument.of(resolvedArgType));
         }
       }
 
       final UdfFactory holder = functionRegistry.getUdfFactory(functionName);
-      final KsqlScalarFunction function = holder.getFunction(argumentTypes);
+      final KsqlScalarFunction function = holder.getFunction(argTypesForFirstPass);
+
+      final List<ParamType> paramTypes = function.parameters();
```

Comment:
Should we check if `function.parameters().size() == arguments.size()` first?

Reply:
I don't think we need to since this would have been already checked in the UdfIndex as part of verifying the candidates with resolveGenerics.